File: matlab/tools/channel_processing/get_elec_locs.m
function output = get_elec_locs(chLabels,filename)
% Get electrode locations for specified channel labels from a file.

% Parameters:
% - chLabels (cell array): Channel labels for which to retrieve electrode locations.
% - filename (string, char): Path to the file containing electrode locations.

% Returns:
% - output (matrix): Matrix containing electrode locations corresponds to channel labels.
%                        Unavailable channel locations filled with nan.

% Example:
% chLabels = {'Fp1', 'Fp2', 'C3', 'C4'};
% filename = 'electrode_locations.csv';
% output = get_elec_locs(chLabels, filename);

p = inputParser;
addRequired(p, 'chLabels', @(x) iscell(x) || isstring(x) || ischar(x));
addRequired(p, 'filename', @(x) (isstring(x) || ischar(x)) && exist(x,'file') == 2);
parse(p, chLabels, filename);
chLabels = p.Results.chLabels;
filename = p.Results.filename;

if ~iscell(chLabels)
    try
        chLabels = cellstr(chLabels); 
    catch ME
        throw(MException('CNTtools:invalidInputType','Electrode labels should be a cell array.'))
    end
end

elec_locs = table2cell(readtable(filename));
% clean both
labels = clean_labels(elec_locs(:,1));
chLabels = clean_labels(chLabels);

[Lia,locb] = ismember(chLabels,labels);
output = nan(size(chLabels,1),size(elec_locs,2)-1);
output(Lia,:) = cell2mat(elec_locs(locb(Lia),2:end));
% output = chLabels;
% output = [output,num2cell(output_data)];

File: python/CNTtools/tools/get_elec_locs.py
import pandas as pd
import numpy as np
from .clean_labels import clean_labels
from beartype import beartype
from beartype.typing import Iterable, Union

@beartype
def get_elec_locs(chLabels: Union[Iterable[str], str], filename: str) -> np.ndarray:
    """
    Get electrode locations for specified channel labels from a file.

    Parameters:
    - chLabels (iterable or str): Channel labels for which to retrieve electrode locations.
    - filename (str): Path to the file containing electrode locations.

    Returns:
    - output (np.ndarray): Numpy array containing electrode locations corresponding to channel labels
                    Unavailable channel locations filled with nan.

    Example:
    chLabels = ['Fp1', 'Fp2', 'C3', 'C4']
    filename = 'electrode_locations.csv'
    output = get_elec_locs(chLabels, filename)
    """

    # Convert single string to list
    if isinstance(chLabels, str):
        chLabels = [chLabels]

    # Load electrode locations from the specified file
    elec_locs = pd.read_csv(filename,header = None,index_col = 0)
    locs = elec_locs.to_numpy()
    # Clean labels from both sources
    labels = clean_labels(elec_locs.index.to_numpy())
    chLabels = clean_labels(chLabels)

    # Find common labels and their corresponding indices
    common = np.isin(chLabels,labels)
    out_locs = np.nan*np.zeros([len(chLabels),locs.shape[1]])
    for i in range(len(chLabels)):
        if common[i]:
            ind = np.where(labels == chLabels[i])[0]
            out_locs[i,:] = locs[ind,:]

    return out_locs
